python/pyarrow/_array.pyx

```diff
@@ -992,6 +993,23 @@ cdef class FixedSizeBinaryValue(ArrayValue):
         return cp.PyBytes_FromStringAndSize(data, length)
 
 
+cdef class StructValue(ArrayValue):
+    def as_py(self):
+        cdef:
+            CStructArray* ap
+            vector[shared_ptr[CField]] child_fields = self.type.type.children()
+        ap = <CStructArray*> self.sp_array.get()
+        child_arrays = ap.fields()
+        boxed_arrays = (box_array(child) for child in child_arrays)
+        child_names = (child.get().name() for child in child_fields)
+        # Return the struct as a dict
+        return {
+            name: value for name, value in (
+                (name, child_array[self.index].as_py())
+                for (name, child_array) in
+                izip(child_names, boxed_arrays)
+            ) if value is not None
```

Comment:
I've meant consistent with what's happening in the new code in pandas_convert.
(note the `if (!arr->field(field_idx)->IsNull(i)) {`).

I realize that the overhead isn't great, but its there. 
I also didn't think of the dict matching the schema as being important since I was thinking in the line of dicts being a gap-stopper solution.
I don't have a strong opinion about this so i'll just make the change :)